presto-clp/src/main/java/com/facebook/presto/plugin/clp/metadata/filter/ClpMetadataFilterProvider.java

```diff
@@ -0,0 +1,175 @@
+            filterMap = mapper.readValue(
+                    new File(config.getMetadataFilterConfig()),
+                    new TypeReference<Map<String, List<ClpMetadataFilter>>>() {});
+        }
+        catch (IOException e) {
+            throw new PrestoException(CLP_METADATA_FILTER_CONFIG_NOT_FOUND, "Failed to open metadata filter config file", e);
+        }
+    }
+
+    /**
+     * Rewrites the given pushed-down expression for metadata filtering to remap filter conditions
+     * based on the configured range mappings for the given scope.
+     * <p></p>
+     * The {@code scope} follows the format {@code catalog[.schema][.table]}, and determines
+     * which filter mappings to apply, since mappings from more specific scopes (e.g., table-level)
+     * override or supplement those from broader scopes (e.g., catalog-level). For each scope
+     * (catalog, schema, table), this method collects all range mappings defined in the metadata
+     * filter configuration.
+     *
+     * @param scope the scope of the filter
+     * @param pushDownExpression the pushed-down expression for metadata filtering that needs to
+     *                           be rewritten
+     * @return the rewritten pushed-down expression for metadata filtering
+     */
+    public abstract String remapMetadataFilterPushDown(String scope, String pushDownExpression);
+
+    /**
+     * Checks for the given table, if the given pushed-down expression for metadata filtering
+     * contains all required fields.
+     *
+     * @param schemaTableName the table that is being queried
+     * @param metadataFilterPushDownExpression the pushed-down expression for metadata filtering
+     *                                         to be checked
+     */
+    public void checkContainsRequiredFilters(SchemaTableName schemaTableName, String metadataFilterPushDownExpression)
+    {
+        boolean hasRequiredMetadataFilterColumns = true;
+        ImmutableList.Builder<String> notFoundListBuilder = ImmutableList.builder();
+        for (String columnName : getRequiredColumnNames(format("%s.%s", CONNECTOR_NAME, schemaTableName))) {
+            if (!metadataFilterPushDownExpression.contains(columnName)) {
+                hasRequiredMetadataFilterColumns = false;
+                notFoundListBuilder.add(columnName);
+            }
+        }
+        if (!hasRequiredMetadataFilterColumns) {
+            throw new PrestoException(
+                    CLP_MANDATORY_METADATA_FILTER_NOT_VALID,
+                    notFoundListBuilder.build() + " is a mandatory metadata filter column but not valid");
+        }
+    }
+
+    public Set<String> getColumnNames(String scope)
+    {
+        return collectColumnNamesFromScopes(scope, this::getAllColumnNamesFromFilters);
+    }
+
+    /**
+     * Returns the {@link MetadataDatabaseSpecific} class implemented by the user. To respect our
+     * code style, we recommend to implementing a {@code protected static class} as an inner class
```

Comment:
```suggestion
     * code style, we recommend implementing a {@code protected static class} as an inner class
```
Sorry, just a small grammatical change